demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs

```diff
@@ -0,0 +1,1033 @@
+            },
+        )?;
+
+        let mut scb_validators_keys_root = NULL_LEAF_G.clone(); //CR:TODO: check if we need other two custom fields for this root value as for scb_new_mst_root
+
+        // Alloc custom_fields and enforce their hash, if they are present
+        let custom_fields_hash_g = if let Some(custom_fields) = self.custom_fields.clone() {
+            let custom_fields_g =
+                Vec::<FieldElementGadget>::alloc(cs.ns(|| "alloc custom fields"), || {
+                    Ok(custom_fields.as_slice())
+                })?;
+
+            assert!(custom_fields_g.len() >= 1);
+
+            //take the validator keys root from the first custom field
+            scb_validators_keys_root = custom_fields_g[0].clone();
+
+            let custom_fields_hash_g = FieldHashGadget::enforce_hash_constant_length(
+                cs.ns(|| "H(custom_fields)"),
+                custom_fields_g.as_slice(),
+            )?;
+            Some(custom_fields_hash_g)
+        } else {
+            None
+        };
+
+        // Alloc btr_fee and ft_min_amount
+        let btr_fee_g = UInt64::alloc(cs.ns(|| "alloc btr_fee"), self.btr_fee)?;
+
+        let ft_min_amount_g = UInt64::alloc(cs.ns(|| "alloc ft_min_amount"), self.ft_min_amount)?;
+
+        // Pack them into a single field element
+        let fees_bits = {
+            let mut bits = btr_fee_g.to_bits_le();
+            bits.reverse();
+
+            let mut ft_min_amount_bits = ft_min_amount_g.to_bits_le();
+            ft_min_amount_bits.reverse();
+
+            bits.append(&mut ft_min_amount_bits);
+            bits
+        };
+
+        let fees_g = FieldElementGadget::from_bits(
+            cs.ns(|| "pack(btr_fee, ft_min_amount)"),
+            fees_bits.as_slice(),
+        )?;
+
+        let quality_g = FieldElementGadget::alloc(cs.ns(|| "alloc quality"), || {
+            self.threshold.ok_or(SynthesisError::Other("AssignmentMissing threshold".to_string()))
+        })?;
+
+        //Reconstruct message as H(sc_id, epoch_number, bt_root, end_cumulative_sc_tx_comm_tree_root, btr_fee, ft_min_amount)
+        let cert_data_hash_g = {
+            let wcert_sysdata_hash_g = FieldHashGadget::enforce_hash_constant_length(
+                cs.ns(|| "H(sc_id, epoch_number, bt_root, valid_sigs, end_cumulative_sc_tx_comm_tree_root, btr_fee, ft_min_amount)"),
+                &[sc_id_g.clone(), epoch_number_g.clone(), mr_bt_g.clone(), quality_g.clone(), end_cumulative_sc_tx_comm_tree_root_g.clone(), fees_g.clone()],
+            )?;
+
+            let preimage = if custom_fields_hash_g.is_some() {
+                vec![custom_fields_hash_g.clone().unwrap(), wcert_sysdata_hash_g]
+            } else {
+                vec![wcert_sysdata_hash_g]
+            };
+
+            FieldHashGadget::enforce_hash_constant_length(
+                cs.ns(|| "H([custom_fields], cert_data_hash)"),
+                preimage.as_slice(),
+            )
+        }?;
```

Comment:
And [this](https://github.com/HorizenOfficial/zendoo-sc-cryptolib/blob/master/demo-circuit/src/blaze_csw/constraints/data_structures.rs#L149)